[18] Updated Suspense doc with behavior during SSR and Hydration (#4484)
reactjs/reactjs.org: content/docs/reference-react.md

While this is not supported today, in the future we plan to let `Suspense` handle more scenarios such as data fetching. You can read about this in [our roadmap](/blog/2018/11/27/react-16-roadmap.html).

#### `React.Suspense` in SSR
During SSR trees wrapped in Suspense are rendered synchronously as deeply as possible. If the boundary never suspends 
then the tree renders as though the boundary were never there. If a component does suspend then we schedule a low 
priority task to render the fallback instead. If the component unsuspends before we flush the fallback then we send 
down the actual content and throw away the fallback.

#### `React.Suspense` during hydration

> Suspense boundaries depend on their parent boundaries being hydrated before they can hydrate, but they can hydrate independently from sibling boundaries. Events on a boundary before its hydrated will cause the boundary to hydrate at 

Reviewer: I was struggling with the "they" references. Does it sound clearer to split this sentence up into two? 
```suggestion
Suspense boundaries depend on their parent boundaries being hydrated before they can hydrate. Suspense boundaries can hydrate independently from sibling boundaries. Events on a boundary before it's hydrated will cause the boundary to hydrate at 
```

While this is not supported today, in the future we plan to let `Suspense` handle more scenarios such as data fetching. You can read about this in [our roadmap](/blog/2018/11/27/react-16-roadmap.html).

#### `React.Suspense` in SSR
During SSR trees wrapped in Suspense are rendered synchronously as deeply as possible. If the boundary never suspends 
then the tree renders as though the boundary were never there. If a component does suspend then we schedule a low 
priority task to render the fallback instead. If the component unsuspends before we flush the fallback then we send 
down the actual content and throw away the fallback.

#### `React.Suspense` during hydration
Suspense boundaries depend on their parent boundaries being hydrated before they can hydrate, but they can hydrate independently from sibling boundaries. Events on a boundary before its hydrated will cause the boundary to hydrate at 
a higher priority than neighboring boundaries but discrete events (eg. click) will not be replayed upon hydration. Only 

> the last continuous event (`focusin`, `mousein`, `dragenter`, `mouseover`, `pointerover`, `gotpointercapture`, `hover`) 

Reviewer: I believe you meant `mouseenter` not `mousein`?

Is `focusin` a continious event?

I think it was brought up earlier whether discrete vs continous events needs a separate section. If this is something devs should be aware of, then it's probably better to link to a list of discrete and continous events.

Because intuitively a `gotpointercapture` is a discrete thing because a pointer being captured is a discrete state from my understanding (pointer is captured or not). Same for `dragenter` I guess. For `mouseover` it makes sense to consider it continous. 

Either way, I think this illustrates that it isn't clear what events are considered "discrete" and which are "continious".

Reviewer 2: This is why I don't like "discrete" as a word. Our usage of "discrete" is "user _meant_ to do this, and there could be a mistake if we batch it with another same event". Our usage of "continuous" is "user _did not_ intentionally mean to do _this particular_ event, and so we can skip/batch it up with next event of the same type". At least to my understanding.

Reviewer 3: It doesn't really have anything to do with the user intent. That's just something we use as an explanation.

Btw, I don't think we should use the term replay. Because actually, we probably shouldn't actually replay the same event. It's more about reissuing an observation of the current state.

This is like if something is animating below your cursor. Even if you don't move the mouse it should ideally trigger mouseenter/leave anyway as the thing below moves in and out to trigger hover effects. CSS hovers do this.

Focus is a little special since when it's accompanied by other events like key events, it effectively becomes discrete as a result of how it's controlled.

Reviewer 3: For purposes of the hydration, these are not the same thing as the "discrete" and "continuous" hints that we use to inform priorities and batching.

I think we shouldn't couple them and probably not use the same terminology.

These events are things where you can issue new events to inform what the state of the world is. If the user twitched a little, it's the same enter/leave happening on a moving target. The goal in the end is to inform what the current state of the world is.

`focusin` is like saying "activeElement is now on this". So restating that "activeElement is now on this" is fine.

Similarly, `gotpointercapture` is like restating that you currently has the pointer captured.

Reviewer 3: There's no `hover` event is there?

Author: We emulate it off `mousenter`

Reviewer: "persistent events" makes sense to me now. Thanks for clarifying!

Reviewer 3: @salazarm Is that an FB-only thing?

Author: @sebmarkbage oops it was a mistake. I updated the events
